[Gradle Jdk Automanagement] Generate Gradle Jdks Configuration & setup Intellij

crogoz opened the pull request:
## Before this PR
<!-- What's wrong with the current state of the world and why change it now? -->

## After this PR
<!-- User-facing outcomes this PR delivers go below -->
Part of: https://github.com/palantir/gradle-jdks/issues/333

- Tasks to generating and check the "gradle/" configurations files required for Gradle JDK automanagement 
- Removing the dependency on baseline-java-versions. The `gradle-jdks` plugin will not apply anymore by default the `baseline-java-versions` plugin, allowing the Java toolcahins setup as explained in: https://docs.gradle.org/current/userguide/toolchains.html
- Setting up Intelij startup tasks to configure the Gradle Jdks setup
- Small fixes to the gradle-jdk-setup.sh script
- added Gradle-jdks-setup/README.md

==COMMIT_MSG==
[Gradle Jdk Automanagement] Generate Gradle Jdks Configuration
==COMMIT_MSG==

#### FLUPS
* Disabling the worklfow
  * https://github.com/palantir/gradle-jdks/pull/360/files?short_path=292baa3#r1619100753
  * https://github.com/palantir/gradle-jdks/pull/360#discussion_r1619058804
* Allow changing the gradle daemon version if the current daemon version is incompatible.
* Windows Setup
* Visual Studio setup


## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

crogoz committed:
wip working

crogoz committed:
address some comments

crogoz committed:
discover app_home & rm insecure curl

crogoz committed:
reuse logic

crogoz committed:
move integration tests & refactor

crogoz committed:
fix & test

crogoz committed:
fix

crogoz committed:
fix certificate

crogoz committed:
local test & blocking locks

crogoz committed:
use java instead to move jdk & set up certs

crogoz committed:
renames & gradle_user_home & configurable certs

crogoz committed:
testing on ubuntu

crogoz committed:
log nonExisting certs

crogoz committed:
cleanup

crogoz committed:
wip working locally

crogoz committed:
fixing tests

crogoz committed:
only when run locally

crogoz committed:
cleanup

crogoz committed:
fix checkstyle

svc-changelog committed:
Add generated changelog entries

crogoz committed:
address comments

crogoz committed:
fixes

crogoz committed:
fix

crogoz committed:
template & fix

crogoz committed:
fix ldd

crogoz committed:
test & match CurrentOs

crogoz committed:
test & match CurrentOs

crogoz committed:
wip

crogoz committed:
test

crogoz committed:
fix test names

crogoz committed:
Merge branch 'cr/gradle-jdk-script' into cr/wrapper-patching

crogoz committed:
wip

crogoz committed:
wip

crogoz committed:
classpaths ./gradlew test

crogoz committed:
load classpath

crogoz committed:
.

crogoz committed:
fix test

crogoz committed:
split

crogoz committed:
move integration tests

crogoz committed:
tests

svc-excavator-bot committed:
Excavator:  Upgrades Baseline to the latest version (#340)

.

crogoz committed:
address comments

crogoz committed:
Merge branch 'develop' into cr/wrapper-patching

crogoz committed:
Excavator:  Upgrade dependencies (#339)

crogoz committed:
Merge branch 'cr/wrapper-patching' of github.com:palantir/gradle-jdks into cr/wrapper-patching

crogoz committed:
test works

crogoz committed:
Merge branch 'develop' into cr/patching2

crogoz committed:
test works

crogoz committed:
.

crogoz committed:
fix tests

crogoz committed:
split gradle-jdks-commons & 2 step patched jar

crogoz committed:
all jdks

crogoz committed:
fix tests

crogoz committed:
versions

crogoz committed:
Merge branch 'develop' into cr/patching2

crogoz committed:
tests toolchains

crogoz committed:
comment

crogoz committed:
message not present all the time

crogoz committed:
messages based on gradle versions

crogoz committed:
local gradle-baseline bump

crogoz committed:
no windows

crogoz committed:
fix

crogoz committed:
fix

crogoz committed:
jdk extension created

crogoz committed:
.

crogoz committed:
bump gradle-jdks

crogoz committed:
fix test

crogoz committed:
changelog

crogoz committed:
Merge branch 'develop' into cr/patching2

crogoz committed:
wip

crogoz committed:
wip dups

crogoz committed:
wip

crogoz committed:
rename gradle-daemon-jdk-version

crogoz committed:
address comments

crogoz committed:
move the jar to gradle dir

crogoz committed:
Merge branch 'cr/patching2' into cr/patchin2-generate-gradle

crogoz committed:
options

crogoz committed:
fix

crogoz committed:
try

crogoz committed:
wip

crogoz committed:
wip

crogoz committed:
.

crogoz committed:
Merge branch 'develop' into cr/patchin2-generate-gradle

crogoz committed:
fix

fix

fixes

fixes

fix test

test

wrapper

wrapper

check

test

.

.

crogoz committed:
fix

crogoz committed:
.

crogoz committed:
add test

CRogers commented on gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/CaResources.java:
```diff
@@ -271,4 +272,15 @@ private static String encodeCertificate(Certificate palantirCert) {
+            X509Certificate cert = (X509Certificate) certFactory.generateCertificate(in);
+            return cert.getSerialNumber().toString();
+        } catch (CertificateException e) {
+            throw new RuntimeException(e);
```
Best to add "could not get serial number for certificate" + `certContent` so we can more easily debug.

CRogers commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -132,16 +135,29 @@ for dir in "$APP_GRADLE_DIR"/jdks/*/; do
     if command -v curl > /dev/null 2>&1; then
       echo "Using curl to download $distribution_url"
-      curl -C - "$distribution_url" | tar -xzf -
+      #
```
leftover comment

CRogers commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -132,16 +135,29 @@ for dir in "$APP_GRADLE_DIR"/jdks/*/; do
       echo "Using curl to download $distribution_url"
-      curl -C - "$distribution_url" | tar -xzf -
+      #
+      if echo "$distribution_url" | grep -q ".zip"; then
```
Can probably use a `case` statement to avoid using `grep`:

```
case "$distribution_url" in
  *.zip)
    # ...
    ;;
  *)
    # ...
    ;;
esac
```

CRogers commented on gradle-jdks/build.gradle:
```diff
@@ -68,5 +69,12 @@ tasks.withType(JavaCompile).configureEach {
+}
+
+tasks.withType(ProcessResources.class).configureEach {
     dependsOn ':gradle-jdks-setup:fatJar'
+    from("${project(':gradle-jdks-setup').projectDir}/build/libs/gradle-jdks-setup-all-${project.version}.jar") {
```
I think you can avoid the dependsOn on by doing `from(':gradle-jdks-setup:fatJar')` (or `from(project(':gradle-jdks-setup').tasks.named('fatJar') if that doesn't work - this one might require an `evaluationDependsOn `gradle-jdks-setup`)

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/FileUtils.java:
```diff
@@ -0,0 +1,82 @@
+                throw exceptionIfFailed;
+            }
+        } catch (IOException e) {
+            throw exceptionIfFailed;
```
Would make this a `Supplier<ExceptionWithSuggestion>` as otherwise this line will be left out of the stacktrace.

CRogers replied:
It feels like a read error (other than file not exists) should not throw the "regenerate" exception here? I would do an explicit check for if the file exists and just have this as a generic `new RuntimeException("failed to read", e)` error.

crogoz replied:
This exception would also account for files that are not present, I can add a check for it. However, if I can't read the file, shouldn't that mean we need to regenerate the files ?

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/FileUtils.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @VisibleForTesting
+    static void checkFilesAreTheSame(File originalPath, File outputPath) {
```
Feels like this could be combined with the above method?

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/FileUtils.java:
```diff
@@ -0,0 +1,82 @@
+                }
+            });
+        } catch (IOException e) {
+            throw exceptionIfFailed;
```
Again this feels odd.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -45,6 +49,9 @@ public abstract class GradleWrapperPatcher {
     private static final String GRADLEW_PATCH = "gradlew-patch.sh";
     private static final String COMMENT_BLOCK = "###";
     private static final String SHEBANG = "#!";
+    private static final ExceptionWithSuggestion REGENERATE_WRAPPER_PATCHER = new ExceptionWithSuggestion(
```
This exception will have a fixed stacktrace based on when this class was created (will be at `GradleWrapperPatcher.<init>`) for this no matter which codepath actual happens when this stacktrace is thrown. Exceptions need to be made at the last possible moment to get accurate info.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -53,14 +60,17 @@ public abstract class GradleWrapperPatcher {
 
         @InputFile
-        RegularFileProperty getOriginalGradlewScript();
+        Property<File> getOriginalGradlewScript();
```
Why `RegularFileProperty` -> `Property<File>`?

crogoz replied:
I switched to using the outputs from the WrapperTask (instead of hardcoding the paths): https://github.com/palantir/gradle-jdks/pull/360/files#diff-d14af7703cef8b606999ab3937204e21f209c24a65a6196c2c63b24ac0fd8c40R111-R112 which are of `Property<File>` instead of `RegularFileProperty`

CRogers replied:
You can use `.fileProvider(` rather than `.set(`.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksExtension.java:
```diff
@@ -54,8 +57,19 @@ public JdksExtension() {
+    }
+
+    public final void setDaemonTarget(String value) {
+        daemonTarget.set(ChosenJavaVersion.fromString(value).javaLanguageVersion());
```
I'd avoid using baseline's `ChosenJavaVersion` here (we don't need the preview functionality) - might as well just use `JavaLanguageVersion.of`.

crogoz committed:
address comments

CRogers commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,169 @@
+        }
+
+        when:
+        def secondCheck = runGradlewTasks('check', '--info')
```
secondCheck runs first?

CRogers commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,169 @@
+                }
+            }
+        }
+        Files.exists(projectDir.toPath().resolve("gradle/gradle-daemon-jdk-version"))
```
This sort of `fileExists` testing feels like it's checking internal state (the layout of files on disk) rather than the actual behaviour we want (Gradle runs with the JDKs we configured). It doesn't really matter what the files in the repo look like - we only really care that this integrates with Gradle properly and our custom JDKs are used at the right time.

I'd at least make sure to have some kind of test that generates the configs then runs a task that actually checks the `System.getProperty('java.home')` of the daemon. Also run Test/JavaCompile tasks and check the output has expected JDK location (or runs a test that writes the java home to a file). In my eyes this would be a proper ETE test.

Similarly, tests that check the config generation tasks are up-to-date but do not actually check all the minutiae the config, and that it errors out when the config files are deleted or the JDK config in gradle is changed.

It might be worth having a test that checks the locations of everything to aid debugging or stuff that can't be tested by running real jdks on circle (like checking macos etc files are generated) - but these seem less important than the overall ETE.

crogoz replied:
I have 2 types of ETE tests: this one that only checks that the gradle configs are generated correctly and the ETE tests in: https://github.com/palantir/gradle-jdks/pull/360/files/87811f803f1e60dc26c8881e8c4bcc626616c99a#diff-98ba650ac01c560cdf5ffae45e4a36b0f4786c3956183b2d9cc1d9759b3061ceR117 (the "javaToolchains" tests the run, javaCompile, daemon toolchains - I didn't have the "Test" task here, will add it shortly)

CRogers replied:
Unfortunately Github doesn't show these diff links properly half the time - what's the test class name?

crogoz replied:
I moved the toolchain ETE test here: [GradleJdkToolchainsIntegrationTest.groovy](https://github.com/palantir/gradle-jdks/pull/360/files#diff-d8cb60820e935e74203f71da27d54ca1e390a3b5cb379b2a15f807309a3a4018)
I'll add another test that doesn't apply the `baseline-java-versions` plugin

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,145 @@
+import org.gradle.api.tasks.wrapper.Wrapper;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+public final class ToolchainsPlugin extends JdkDistributionConfigurator implements Plugin<Project> {
```
Feels weird to extend `JdkDistributionConfigurator` when it just has static methods - I'd just call the statics as normal.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksPlugin.java:
```diff
@@ -42,76 +37,21 @@ public void apply(Project rootProject) {
         }
         rootProject.getPluginManager().apply(BaselineJavaVersions.class);
```
I think we only need to apply `BaselineJavaVersions` if it's the `BaselineJavaJdksPlugin` now? Could even just apply it there.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleJdkConfigs.java:
```diff
@@ -0,0 +1,237 @@
+
+    private static final Logger log = Logging.getLogger(GradleJdkConfigs.class);
+    private static final String JDKS_DIR = "jdks";
+    private static final String CERTS_DIR = "certs";
+    private static final String DOWNLOAD_URL = "download-url";
+    private static final String LOCAL_PATH = "local-path";
```
For constants that are used in exactly one place I think it's better to just inline them - it's only worth making a static field when it's used in multiple places that must be the same.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -75,41 +87,51 @@ interface Params {
-                    .map(setupJar -> setupJar.getAsFile().exists())
-                    .getOrElse(false));
+            getGenerate().convention(false);
+            onlyIf(t -> getGradleJdksSetupJar().map(File::exists).getOrElse(false));
```
Do we need this onlyIf? Can we ever end up running this task without extracting the gradle-jdks-setup jar first?

CRogers replied:
Just worried it's going to silently do nothing rather than explode in an obvious way.

crogoz replied:
> Do we need this onlyIf? Can we ever end up running this task without extracting the gradle-jdks-setup jar first?

You are right, this shouldn't happen.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ProjectToolchainsPlugin.java:
```diff
@@ -0,0 +1,57 @@
+        JdkDistributions jdkDistributions = new JdkDistributions();
+        JdksExtension jdksExtension = project.getRootProject().getExtensions().getByType(JdksExtension.class);
+
+        project.getPluginManager().apply(BaselineJavaVersion.class);
```
This should be a `withPlugin` as we want `baseline-java-versions` to be optional in the new plugin.

crogoz committed:
fixes & address comments

crogoz committed:
fix

crogoz committed:
baseline-java-versions optional

crogoz committed:
fix test with baseline-java

crogoz committed:
javaVersions loaded from jdks

crogoz committed:
refactor tests

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,155 @@
+                .getTasks()
+                .named(LifecycleBasePlugin.CHECK_TASK_NAME)
+                .configure(check -> check.dependsOn(checkGradleJdkConfigs, checkWrapperPatcher));
+
```
Pretty sure we're going to want a single `setupJdks` task or something that depends on the two generate tasks so we can call just one task in the rollout excavator.

CRogers commented:
:+1:

crogoz committed:
wip

crogoz committed:
fix up-to-date checks

crogoz committed:
small cleanups

crogoz committed:
refactor

crogoz committed:
fix checkstyle + add javatoolchains test

crogoz commented on gradle-jdks-setup/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:
```diff
@@ -91,7 +91,7 @@ private static List<String> getAndConfigureJdkAutomanagement() {
             Path gradleJdksConfigurationPath, Path gradleJdksInstallationDir, String osName, String archName) {
         try (Stream<Path> gradleJdkConfigurationPath =
-                Files.walk(gradleJdksConfigurationPath, 1).filter(Files::isDirectory)) {
+                Files.list(gradleJdksConfigurationPath).filter(Files::isDirectory)) {
```
small fix - `.walk` will also include the current directory, but I need to iterate through the subDirectories.

crogoz commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -84,7 +84,7 @@ read_value() {
     die "ERROR: $1 not found, aborting Gradle JDK setup"
   fi
-  read -r value < "$1"
+  read -r value < "$1" || die "ERROR: Unable to read value from $1. Make sure the file ends with a newline."
```
reads the configuration files (eg. `gradle/jdks/17/macos/aarch64/download-url`)
`read` expects to have an endline at the end, otherwise it will return a non-zero exit code and because we have `set -o pipefail` enabled (see line#42), the script would fail silently

crogoz commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -132,16 +132,34 @@ for dir in "$APP_GRADLE_DIR"/jdks/*/; do
     if command -v curl > /dev/null 2>&1; then
       echo "Using curl to download $distribution_url"
-      curl -C - "$distribution_url" | tar -xzf -
+      case "$distribution_url" in
```
special-casing `zip` files.  I can't download and extract zip files like before (`curl -C - "$distribution_url" | tar -xzf -`) because the file permissions will not be maintained. Doing this in 2 steps maintains the permissions.

felixdesouza replied:
I did not realise you could untar a zip file 🤯

crogoz commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -158,6 +176,7 @@ done
 rm -rf "$tmp_work_dir"
 
+# TODO(crogoz): what happens if gradle-daemon-jdk-version doesn't exist or the jdk for that version doesn't exist ?
 gradle_daemon_jdk_version=$(read_value "$APP_GRADLE_DIR"/gradle-daemon-jdk-version)
```
if `gradle-daemon-jdk-version` is not present or the JDK version specified does not exist in gradle/jdks (https://github.com/palantir/gradle-jdks/pull/360/files#diff-b43a6a48890fa88a6f76f47d3b3ac00c70432c5451f9551ee0d81e1f5db6ec7eR181), then `./gradlew task` will fail. I'll add a FLUP PR for these cases.

crogoz commented on gradle-jdks/build.gradle:
```diff
@@ -68,5 +69,11 @@ tasks.withType(JavaCompile).configureEach {
+    systemProperty 'ignoreDeprecations', 'true'
+}
+
+tasks.withType(ProcessResources.class).configureEach {
```
`gradle-jdks-setup.jar` needs to be added to the `gradle/` configuration directory when the JDK setup is enabled. It is used by the patched  `./gradlew` [to run the JDK setup](https://github.com/palantir/gradle-jdks/blob/2f1feae30cad4266dad85d29ba22a47ddaa59a29/gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh#L154)

The jar will be included in the `gradle-jdks` resources, such that when the `gradle-jdks` plugin is applied, when the  task `wrapperPatcherTask` is called, we can download the jar from the resources and copy it in the 'gradle/' directory. (https://github.com/palantir/gradle-jdks/pull/360/files#diff-664db41a37fb84c50fd2cf1579ff8a17040c6becc39198bb87ecdb468675ddf7R73)

felixdesouza replied:
This will add it to every source set in this project. Is there any way to apply it to just the main source set?

CRogers replied:
adding `.named('processResources')` would do it. Or more generally:

```java
project.getTasks()
        .withType(ProcessResources.class)
        .named(project.getExtensions()
                .getByType(SourceSetContainer.class)
                .getByName(SourceSet.MAIN_SOURCE_SET_NAME)
                .getProcessResourcesTaskName());
```

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/BaselineJavaJdksPlugin.java:
```diff
@@ -0,0 +1,98 @@
+import org.gradle.jvm.toolchain.JavaInstallationMetadata;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+
+public final class BaselineJavaJdksPlugin implements Plugin<Project> {
```
Moved the current implementation from [`JdksPlugin`](https://github.com/palantir/gradle-jdks/pull/360/files#diff-9ca2e698760ae7c328bed1a9e7136fbe457083f7775fb90a0b870ffa7c532afaL43). This is just a copy-paste.

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleJdkConfigs.java:
```diff
@@ -0,0 +1,92 @@
+ * The two tasks {@link CheckGradleJdkConfigsTask} and {@link GenerateGradleJdkConfigsTask} need to either check the
+ * validity of the already generated gradle files or generate the gradle files.
+ */
+public abstract class GradleJdkConfigs extends DefaultTask {
```
both `CheckGradleJdkConfigsTask` and `GenerateGradleJdkConfigsTask` need the same logic to check or generate files (see: https://github.com/palantir/gradle-jdks/pull/360/files#diff-76cb1e65229cb1f4a54d1ddbeb74d90ffe95842c31be1faae522544da1e41040R64-R91). I've tried to move the common logic to this class, then each of the other classes will deal with implementing the specific actions (check/generation) for each of the files (see: https://github.com/palantir/gradle-jdks/pull/360/files#diff-76cb1e65229cb1f4a54d1ddbeb74d90ffe95842c31be1faae522544da1e41040R52-R61)

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleJdkConfigs.java:
```diff
@@ -0,0 +1,92 @@
+    @Input
+    public abstract MapProperty<String, String> getCaCerts();
+
+    public abstract Directory gradleDirectory();
```
for `GenerateGradleJdkConfigsTask ` the `gradleDirectory` is an `@Output`, for `CheckGradleJdkConfigsTask ` the `gradleDirectory` is an `@Input`.

crogoz commented on gradle-jdks/src/main/resources/gradlew-patch.sh:
```diff
@@ -1,5 +1,5 @@
-
 # >>> Gradle JDK setup >>>
+# TODO(crogoz): what if we disable the gradle.jdk.setup.enabled=false?
 # !! Contents within this block are managed by 'palantir/gradle-jdks' !!
```
Will add a FLUP - if an env variable is set (or if `gradle.jdk.setup.enabled=false`) -> we won't run the jdks-setup script.

crogoz committed:
fix test

crogoz committed:
readme finished

crogoz committed:
readme

crogoz committed:
bug fix: generate all the jdks from jdksExtension, not only from baseline-java-versions

crogoz committed:
cleanup

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/FileUtils.java:
```diff
@@ -0,0 +1,49 @@
+                        .get();
+            }
+        } catch (IOException e) {
+            throw exceptionIfFailed.map(Supplier::get).orElseThrow(() -> new RuntimeException(e));
```
don't you end up swallowing the exception here? Can you add it as a Function over Supplier? and then either have it as suppressed or something else.

felixdesouza replied:
I'm also not sure where this method is used, I can't seem to see it.

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GenerateGradleJdkConfigsTask.java:
```diff
@@ -0,0 +1,123 @@
+    @Override
+    public final Directory gradleDirectory() {
+        return getOutputGradleDirectory().get();
+    }
```
curious where this is used if it's public? this will force the task to run or it will expect it to have run.

crogoz replied:
you are right, will change it, no need to be public

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+    @Override
+    public void apply(Project rootProject) {
+        if (!JdksPlugin.getEnableGradleJdkProperty(rootProject)) {
+            throw new RuntimeException("Cannot apply ToolchainsJdksPlugin without enabling gradle.jdk.setup.enabled");
```
is this our own property? if so, I think we should avoid prefixing with `gradle` as it looks like it's an official property. Perhaps we can prefix with `palantir` ? Then it should also be a bit more searchable.

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+            rootProject
+                    .getExtensions()
+                    .getByType(BaselineJavaVersionsExtension.class)
+                    .getSetupJdkToolchains()
```
implementation details on gradle-baseline side [here](https://github.com/palantir/gradle-baseline/pull/2783)

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                                    jdkDistributions,
+                                    jdksExtension.getConfiguredJavaVersions().get().stream()
+                                            .filter(javaLanguageVersion ->
+                                                    javaLanguageVersion.canCompileOrRun(MINIMUM_SUPPORTED_JAVA_VERSION))
```
TODO - check if there are repos that required java < 11

felixdesouza commented on gradle-jdks-setup/src/main/java/org/gradle/wrapper/GradleWrapperMain.java:
```diff
@@ -106,12 +106,7 @@ private static Stream<String> getInstallationPath(
-                    osName,
-                    archName,
-                    gradleJdkConfigurationPath.toAbsolutePath()));
+            System.out.printf("Couldn't find a valid JDK installation %s. Skipping ...%n", localPathFile);
```
I imagine you can't really fail here, because you can never fix it by running another gradle task right?

crogoz replied:
Yes, correct

felixdesouza commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -84,7 +84,7 @@ read_value() {
-  read -r value < "$1"
+  read -r value < "$1" || die "ERROR: Unable to read value from $1. Make sure the file ends with a newline."
   echo "$value"
 }
```
I think this was the file I was referring to when I Was like could you have used a graal compiled binary for this. As it's not doing any gradle work.

felixdesouza commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -129,29 +129,46 @@ for dir in "$APP_GRADLE_DIR"/jdks/*/; do
+      case "$distribution_url" in
+        *.zip)
+          distribution_name=${distribution_url##*/}
+          curl -C - "$distribution_url" -o "$distribution_name"
```
I think you'll want to remove the zip file when you're done?

crogoz replied:
yes! the full temp directory (where we are also downloading the zip file) will be deleted at the end, see: https://github.com/palantir/gradle-jdks/blob/develop/gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh#L159

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksExtension.java:
```diff
@@ -54,8 +59,24 @@ public JdksExtension() {
                 MapProperty.class, getObjectFactory().mapProperty(String.class, String.class));
         this.jdkStorageLocation = SynchronizedInterface.synchronizeAllInterfaceMethods(
                 DirectoryProperty.class, getObjectFactory().directoryProperty());
+        this.daemonTarget = project.getObjects().property(JavaLanguageVersion.class);
```
don't think you need to inject `project` here if you're only using `getObjects()`. It should be sufficient to use the `getObjectFactory()` method if you're not using `project` for anything else.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,145 @@
+            Files.readString(certFile).trim() == "18126334688741185161"
+        } else {
+            !Files.exists(certFile)
+        }
```
I think we should just generate a new cert and use that? It's not specific to the root ca right?

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,145 @@
+        Files.exists(projectDir.toPath().resolve("gradle/gradle-daemon-jdk-version"))
+        Path jarInProject = projectDir.toPath().resolve("gradle/gradle-jdks-setup.jar");
+        Path originalJar = Path.of("build/resources/main/gradle-jdks-setup.jar");
+        Files.exists(jarInProject)
```
I think we can get rid of the `Files.exists` checks, especially if you're going to compare the content later on => existence + will fail when it doesn't exist.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkIntegrationTest.groovy:
```diff
@@ -0,0 +1,193 @@
+            jdks {
+               jdk(11) {
+                  distribution = 'azul-zulu'
+                  jdkVersion = '11.54.25-11.0.14.1'
```
l think we should add parameters to this method to make it clearer where something is coming from:

```groovy
def setupJdksHardcodedVersions(Map versions) {
  ...
  jdkVersion = ${versions.java11}
  ...
}
```

Then when you call it:

```groovy
setupJdksHardcodedVersions(java11: '11.54.25-11.0.14.1', ...)
```

Or better still save it to some variable/constant and then use that same variable/constant when asserting the value is correct.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        writeHelloWorld(subprojectLib1)
+        runTasksSuccessfully('wrapper', '--info')
+
+        when:
```
I've typically found that it can be clearer to demonstrate what you're trying to assert by just adding a string after the `when:` block.

i.e. 

```groovy
when: '<doing something>'
...
then: '<what you're trying to show>'
...

when: '<doing something else>'
...
then

felixdesouza replied:
spock describe it as [Specifications as Documentation](https://spockframework.org/spock/docs/2.3/spock_primer.html#specifications_as_documentation)

CRogers replied:
I have also started doing this

CRogers replied:
(after seeing Felix do it)

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        String runOutput = runGradlewTasksSuccessfully("run", "--info")
+
+        then:
+        assertToolchainsOutput(output, runOutput)
```
let's separate this assert for both the `output` and the `runOutput` into separate `when/then` blocks

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        """.stripIndent(true)
+
+        //language=groovy
+        def subprojectLib = addSubproject 'subprojectLib', '''
```
`subprojectLib` -> `subproject-lib-21`

similarly

`subprojectLib` -> `subproject-lib-11`

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        '''.stripIndent(true)
+        writeHelloWorld(subprojectLib)
+
+        def subprojectLib1 = addSubproject 'subprojectLib1', '''
```
I'm not sure what the purpose of the subprojects here are. You're not checking the output of them in any meaningful way I don't believe?

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        gradleVersionNumber << [GRADLE_7VERSION, GRADLE_8VERSION]
+    }
+
+    def assertToolchainsOutput(String output, String runOutput) {
```
I generally try to avoid mass assertion methods, because it can be quite confusing what's being asserted + hides the flow of stuff i.e. hard coded jdks that are also not configured inside here.

In the same vein of @CRogers' comment, I don't think the exact file name is that important. 

I think you can kill two birds with one stone where:
* you have a `Main` class, that can look at its own `java.home` property. 
* You run this main class and pass it the value of the `paths`  gradle property. 
* Then inside, you want to make sure that the java home is in one of those.

That way, you can probably also disregard having to pass around versions and checking stuff directly, and the test becomes a bunch of main classes which self-report/check that something is expected.

So something like:

```groovy
def '#gradleVersionNumber: gradle-jdks correctly set-up for #mode'() {
        buildFile << """
            java {
                toolchain {
                    languageVersion = JavaLanguageVersion.of($javaVersion)
                }
            }
        """.stripIndent(true)

        file(... + "/Main.java") << """
            public static void main([]String args) {
              String javaHome = System.getProperty("java.home");
              // can either read it as a property or read it as an arg
              String configuredJavaHomes = System.getProperty(...);

             // fail if javaHome not in configuredJavaHomes
            }
        """.stripIndent(true)

        buildFile << """
          // setup the run task to have the right things based on the mode
        """
        
        // Using spocks data driven testing you can end up doing something like this:   
        where: 
        [gradleVersionNumber, javaVersion, mode] << [
           [GRADLE_7VERSION, GRADLE_8VERSION], 
           [11,17,21], 
           ['toolchain', 'baseline-java']
         ].combinations()
```

More on spock's [data driven testing](https://spockframework.org/spock/docs/2.3/data_driven_testing.html)

felixdesouza replied:
It also turns out that if you move your assertions into a method, you *must* add `assert` on the conditions.

The assertions above `runOutput.contains(...)` are not actually checked at all as a result. This currently works because the return value is optional => the assertion is still happening at the call site, but only for the return value.

See [here](https://spockframework.org/spock/docs/2.3/spock_primer.html#_helper_methods):

> When factoring out conditions into a helper method, two points need to be considered: First, implicit conditions must be turned into explicit conditions with the assert keyword. ...

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        then:
+        assertToolchainsOutput(output, runOutput)
+        File compiledClass = new File(projectDir, "build/classes/java/main/Main.class")
+        assertBytecodeVersion(compiledClass, JAVA_17_BYTECODE, 0)
```
// personal preference, so feel free to ignore
Generally prefer to do stuff with some sort of object and assert on that object, then I can actually see what I'm trying to assert here.

i.e.

```
readByteCodeVersion(compiledClass) == ByteCodeVersion.of(JAVA_17_BYTECODE, 0)
```

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkToolchainsIntegrationTest.groovy:
```diff
@@ -0,0 +1,195 @@
+        file('src/main/java/Main.java') << java17Code
+
+        // language=Groovy
+        buildFile << """
```
I think you should also have a task that prints out or verifies that the daemon version is sufficient. You could parse it out of `./gradlew --info` but just feels like the assertion code becomes noisy.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkWrapperPatcherIntegrationTest.groovy:
```diff
@@ -0,0 +1,132 @@
+        List<String> gradlewPatchRows = Files.readAllLines(Path.of('src/main/resources/gradlew-patch.sh'))
+        rowsAfterPatching.removeAll(initialRows)
+        rowsAfterPatching.removeAll(gradlewPatchRows)
+        rowsAfterPatching.size() == 0
```
Wouldn't it be easier to say:

```groovy
def gradlewBeforePatching = file('gradlew').text

...

def gradlewAfterPatching = file('gradlew').text
def patch = new File('src/main/resources/...').text

gradlewAfterPatching.startsWith(patch)
gradlewAfterPatching.endsWith(patch)
```

This should be functionally identical. `removeAll` will remove multiple copies of the same line, if the aim was to detect being patched multiple times.
```

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkWrapperPatcherIntegrationTest.groovy:
```diff
@@ -0,0 +1,132 @@
+
+        then:
+        !output.standardOutput.contains('wrapperJdkPatcher')
+        !output.standardOutput.contains("Gradle JDK setup is enabled, patching the gradle wrapper files")
```
I think depending on the standard output is quite brittle (you might end up changing the underlying text and the test would not have validated anything). 

If there is a task that you don't expect to be run, are you able to use: `!output.wasExecuted(...)`.

NB: I guess my comment holds for the task name, but if you extract it to a constant, you can then assert that it was run in the one where you expect it to run.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkIntegrationTest.groovy:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    String upgradeGradleWrapper() {
+        return runGradlewTasksSuccessfully("getGradleJavaHomeProp", "wrapper", "--gradle-version", "8.4", "-V")
```
this is quite sneaky, and not obvious that you're running an additional task in here.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkWrapperPatcherIntegrationTest.groovy:
```diff
@@ -0,0 +1,132 @@
+
+        when:
+        String wrapperResult1 = upgradeGradleWrapper()
+        String wrapperResult2 = upgradeGradleWrapper()
```
it's not immediately obvious why you need to upgrade twice or what you're trying to assert here.

I think you should have multiple `when/then` blocks and a nice description of what you're trying to assert.

My guess is that you're testing the finaliser behaviour?

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GradleJdkWrapperPatcherIntegrationTest.groovy:
```diff
@@ -0,0 +1,132 @@
+        wrapperResult1.contains(String.format("Successfully installed JDK distribution in %s", expectedLocalPath))
+        String expectedJdkLog = "JVM:          11.0.14.1 (Azul Systems, Inc. 11.0.14.1+1-LTS)"
+        wrapperResult1.contains(expectedJdkLog)
+        wrapperResult1.contains("Gradle 7.6.4")
```
why does upgrading the wrapper still report the version as 7.6.4?

crogoz replied:
I'll rewrite this bit (it is related to your other comments above). 
The first time I am run `upgradeGradleWrapper` I run:  `wrapper --gradle-version 8.4 *-V*`, which will update the `./gradlew` & `gradle-wrapper.jar` files, but the version is still 7.6.4. the next time I run ./gradlew, the correct version of the wrapper will be picked up.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        where:
+        gradleVersionNumber << [GRADLE_7VERSION, GRADLE_8VERSION]
```
let's make this into a data table:

```groovy
gradleVersionNumber | expectedErrorLines
GRADLE_7VERSION | ["No compatible toolchains found for request specification: {languageVersion=15"]
GRADLE_8VERSION | ["No matching toolchains found for requested specification: {languageVersion=15", "No locally installed toolchains match and toolchain auto-provisioning is not enabled."]
```

Then you can assert inside.

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,145 @@
+
+        then:
+        !checkGradleJdkConfigs.contains("checkGradleJdkConfigs UP-TO-DATE")
+        checkGradleJdkConfigs.contains(String.format("Gradle JDK configuration file `%s` is out of date", jdk17Path))
```
nit, let's use String interpolation:
```
"Gradle JDK configuration file `${jdk17Path}` is out of date"
```

felixdesouza commented on gradle-jdks/src/test/groovy/com/palantir/gradle/jdks/GenerateGradleJdkConfigsIntegrationTest.groovy:
```diff
@@ -0,0 +1,145 @@
+
+        when:
+        String jdk17Path = String.format("gradle/jdks/17/%s/%s/local-path", CurrentOs.get().uiName(), CurrentArch.get().uiName());
+        Files.write(projectDir.toPath().resolve(String.format("gradle/jdks/17/%s/%s/local-path", CurrentOs.get().uiName(), CurrentArch.get().uiName())), "new-path\n".getBytes())
```
String interpolate here as well

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                                    rootProject.getRootDir().toPath().resolve("gradlew")));
+                    task.getPatchedGradleWrapperJar()
+                            .set(rootProject.file(
+                                    rootProject.getRootDir().toPath().resolve("gradle/wrapper/gradle-wrapper.jar")));
```
I don't think this task is ever up to date? since you're modifying the same thing as your input?

felixdesouza replied:
Or at the very least, you have to run it twice before it says up-to-date? (assuming you return early if you have no changes to make?)

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                            .get()
+                            .getPatchedGradleWrapperJar()
+                            .getAsFile()
+                            .get()));
```
I think you should use the same inputs as the `wrapper` task as that's what you want to check right?

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                            .getPatchedGradleWrapperJar()
+                            .getAsFile()
+                            .get()));
+                    task.getBuildDir().set(task.getTemporaryDir());
```
TIL about `task.getTemporaryDir` !

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                            .getOutputGradleDirectory()
+                            .file(GradleJdkConfigs.GRADLE_JDKS_SETUP_JAR)
+                            .get()
+                            .getAsFile()));
```
you should be able to use `generatedGradleJdkConfigs.flatMap(t -> t.getOutputGradleDirectory().getLocationOnly())`

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                                    jdksExtension.getConfiguredJavaVersions().get().stream()
+                                            .filter(javaLanguageVersion ->
+                                                    javaLanguageVersion.canCompileOrRun(MINIMUM_SUPPORTED_JAVA_VERSION))
+                                            .collect(Collectors.toSet()),
```
I've thought about it a bit more, and it seems a bit odd to be able to support using jdk 8 but not be able run gradle with it. If I've understood correctly, this is just the version of the jvm used to run the gradle daemon right? If that's the case, then perhaps we fail explicitly on the daemon java version being set to 8 so that we can't get into a bad state.

crogoz replied:
The failure will happen during the configuration phase: https://github.com/palantir/gradle-jdks/blob/develop/gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh#L154. Basically if I add to the `jdkExtension` java versions < 8, when we will run the line: 
`"$java_home"/bin/java -cp "$APP_GRADLE_DIR"/gradle-jdks-setup.jar com.palantir.gradle.jdks.setup.GradleJdkInstallationSetup` with the `java_home` == version 8, it will fail with `Unsupported class version` because `GradleJdkInstallationSetup` was compiled with java version 11, while the current java version would be 8.

I think I could refactor the script such that this line runs with a different java home (probably use the gradle daemon java version to do the setup for all the jdks that I am downloading). This will involve some refactoring though.

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+                                    jdksExtension)));
+                    task.getCaCerts().putAll(jdksExtension.getCaCerts());
+                    task.getOutputGradleDirectory()
+                            .set(rootProject.getLayout().dir(rootProject.provider(() -> rootProject.file("gradle"))));
```
why do you have to wrap this in a `provider()` ?

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,162 @@
+        });
+        TaskProvider<Wrapper> wrapperTask = rootProject.getTasks().named("wrapper", Wrapper.class);
+
+        TaskProvider<GenerateGradleJdkConfigsTask> generateGradleJdkConfigs = rootProject
```
For this style of task, it might be better to register the tasks separately. Then do a 
```
rootProject.getTasks().withType(GradleJdkConfigs.class).configureEach(task -> {
  // set common properties here
});
```

Then they don't depend on each other. Just whatever they previously depended on, which I think is fine.

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksExtension.java:
```diff
@@ -54,8 +59,24 @@ public JdksExtension() {
         this.getCaCerts().finalizeValueOnRead();
         this.getJdkStorageLocation().finalizeValueOnRead();
+        this.getDaemonTarget().finalizeValueOnRead();
+        this.getConfiguredJavaVersions().finalizeValueOnRead();
```
Could you add this to `LazilyConfiguredMapping` ? to be able to get the keys?

I haven't done enough archaeology to figure out why we need `LazilyConfiguredMapping` however. So curious whether doing what you're doing here similarly triggers the same condition that necessitated creating `LazilyConfiguredMapping` initially.

crogoz replied:
> Could you add this to LazilyConfiguredMapping ? to be able to get the keys?

I think this would defeat the purpose of the lazily configured mapping. The keys/values are computed as soon as someone [asks](https://github.com/palantir/gradle-jdks/blob/78445ecafc147fcb1662be8e283f8f384e65907a/gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksPlugin.java#L68) for a specific [java version](https://github.com/palantir/gradle-utils/blob/develop/lazily-configured-mapping/src/main/java/com/palantir/gradle/utils/lazilyconfiguredmapping/LazilyConfiguredMapping.java#L76). The `LazilyConfiguredMapping` doesn't know the keys. 

I think the purpose of `LazilyConfiguredMapping` was to only load the distribution mappings for the java versions we are actually using.

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -138,26 +172,45 @@ private static void write(Path destPath, String content) {
-                .filter(i -> initialLines.get(i).equals(GRADLEW_PATCH_HEADER))
+    private static List<String> getLinesWithoutPatch(File gradlewFile) {
+        List<String> initialLines = readAllLines(gradlewFile.toPath());
+        Optional<Pair<Integer, Integer>> startEndPatch = getPatchLineNumbers(initialLines);
```
`startEndPatch` -> `patchLineRange`

felixdesouza replied:
You should also just use the `Range` type from guava

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -138,26 +172,45 @@ private static void write(Path destPath, String content) {
+    private static List<String> getPatchedLines(File gradlewFile) {
+        List<String> initialLines = readAllLines(gradlewFile.toPath());
+        Optional<Pair<Integer, Integer>> startEndPatch = getPatchLineNumbers(initialLines);
+        return startEndPatch
```
nit: just inline this and don't both saving `getPatchLineNumbers(...)` to a variable?

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -138,26 +172,45 @@ private static void write(Path destPath, String content) {
-            linesNoPatch.addAll(initialLines.subList(endIndex.getAsInt() + 1, initialLines.size()));
-        }
-        return linesNoPatch;
+        return Optional.of(Pair.of(startIndex.getAsInt(), endIndex.getAsInt()));
```
think you also want to check that end > start?

crogoz replied:
I think that would always be the case - https://github.com/palantir/gradle-jdks/pull/360/files/bfde0722541258ce9b98dc2a35f9d60bdbcdfe8c#diff-b07176f3d698d61919834db279976c62c51454eb78b6fab815ff0beb3b466e3bR206 I am searching from startIndex to the end

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -138,26 +172,45 @@ private static void write(Path destPath, String content) {
-        OptionalInt endIndex = IntStream.range(startIndex.getAsInt(), initialLines.size())
-                .filter(i -> initialLines.get(i).equals(GRADLEW_PATCH_FOOTER))
+        OptionalInt endIndex = IntStream.range(startIndex.getAsInt(), content.size())
+                .filter(i -> content.get(i).equals(GRADLEW_PATCH_FOOTER))
```
not sure whether this is supposed to be the place that checks whether you've double patched, but would you want to check that you only have one?

crogoz replied:
makes sense, I will add a check for that as well

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -75,48 +83,74 @@ interface Params {
+            JarResources.extractJar(gradleJdksSetupJar, gradleJdksExtracted);
+            Path gradleWrapperMainFromSetupJar =
+                    OriginalGradleWrapperMainCreator.getGradleWrapperClassPath(gradleJdksExtracted);
+            FileUtils.checkFilesAreTheSame(gradleWrapperMain.toFile(), gradleWrapperMainFromSetupJar.toFile());
```
I think you want to double check whether or not either of the files exist and give a better error for that.

felixdesouza commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java:
```diff
@@ -75,48 +83,74 @@ interface Params {
+    private static void checkGradleJarContainsPatch(File buildDir, File gradleWrapperJar, File gradleJdksSetupJar) {
+        try {
+            Path gradleWrapperExtractedDir = buildDir.toPath().resolve("check-gradle-wrapper-patcher/gradle-wrapper");
+            Files.createDirectories(gradleWrapperExtractedDir);
```
nit: let's just let `JarResources.extractJar` deal with creating the directory properly. Less noisy.

crogoz committed:
address some comments

crogoz committed:
wip

crogoz committed:
wip

crogoz committed:
wip

crogoz committed:
gradle.properties setup jdk patcher

crogoz committed:
cleanup

crogoz committed:
fixing tests

crogoz committed:
wip

crogoz committed:
fixes

crogoz committed:
cleanup

crogoz committed:
fixes

crogoz committed:
fix java code

crogoz committed:
checkstyle

crogoz committed:
windows

crogoz committed:
Merge branch 'develop' into cr/patchin2-generate-gradle

crogoz committed:
rm todo

crogoz committed:
fixes artifactory url & tests

CRogers commented on gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/IntelijGradleJdkSetup.java:
```diff
@@ -0,0 +1,71 @@
+        if (gradleXml.toFile().exists()) {
+            XmlPatcher.updateGradleJvmValue(gradleXml);
+        } else {
+            throw new RuntimeException(
+                    "Could not update the gradle configuration programmatically. Please configure `Gradle JVM` to "
+                            + " GRADLE_LOCAL_JAVA_HOME in ( Settings | Build, Execution, Deployment | Build Tools | "
+                            + "Gradle | Gradle JVM).");
```
I'm hitting this when trying it out with an "empty" `.idea` (ie running `git clean -xf .idea`). Is this because we haven't run the Gradle import yet so it doesn't exist?

CRogers replied:
Weirdly, from a clean checkout once I hit the this error, I never get the `gradle.xml` to appear, even if I manually change the Gradle JDK in the intellij settings.

crogoz replied:
This will always happen for newly imported projects in Intelij. This happens because we cannot patch `.idea/gradle.xml` because the file was not yet generated by Intelij (it takes a few minutes for Intelij to generate the .idea files after the project is loaded). This would be a one-time-setup the users would need to do. 

(It would have been nice if the log message contained a link to the IDE Settings instead of hardcoding the Settings Path)

crogoz replied:
> gradle.xml to appear, even if I manually change the Gradle JDK in the intellij settings.

What happens if you regenerate the Gradle project ?

CRogers replied:
I think this file is only written out when you close intellij.

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,246 @@
+│   ├── config.properties
+├── .idea/
+│   ├── runConfigurations/
+│   │   ├── IntelijGradleJdkSetup.xml
```
This is currently not gitignored and should be (if you run `git clean -xn .idea` you'll see what would not exist in a clean checkout)

crogoz replied:
yes! I have a fix for this on the way

crogoz committed:
wip

projectActivity & fix tests

.

symlinks try

fix settings

fixes

fixes docker base image

fixes

extra tests

cleanup

cleanup

rm tags

cleanup

wip[

.

crogoz commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,195 @@
+
+        Path gradleJdksLocalDirectory = settings.getRootDir().toPath().resolve("gradle/jdks");
+        if (!Files.exists(gradleJdksLocalDirectory)) {
+            logger.debug("Not setting the Gradle JDK properties because gradle/jdks directory doesn't exist. Please run"
```
we are applying the settings plugin, but if the `setupJdks` has not run yet, we won't have the `gradle/jdks` directory populated yet.

crogoz commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,118 @@
+            setupJdksTask.dependsOn(generateGradleJdkConfigs, wrapperPatcherTask);
+        });
+
+        rootProject.getTasks().named("javaToolchains").configure(task -> {
```
WIP - hacky solution at the moment to make sure that when running the intelij startup task - https://github.com/palantir/gradle-jdks/pull/360/files#diff-4e8f06bc09ead2db0d11b0b160158e0fc79f6d700c6ae08b22dd523a2504263aR91, we are running both the "checkGradleJdkConfigs" and "javaToolchains".

Ideally I'd like to have a lifecycle task `ideaSetup` that triggers both `checkGradleJdkConfigs` and `javaToolchains`. Atm I have some issues retrieving the available gradle tasks in the Intellij plugin to check if a specific task exists. I wouldn't want to depend on the gradle property `palantir.setup.jdk.setup` as this might break when the flag won't be used anymore/the new workflow will be enabled by default.

CRogers replied:
Do we need this anymore now we stopped calling gradle from intellij?

crogoz committed:
intelij works

fix

crogoz committed:
.

.

crogoz committed:
message

crogoz committed:
fixes

crogoz committed:
fix intellij

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/CurrentOs.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Simplified version of CurrentOs.java.
+ * @see <a href="file:../gradle-jdks-setup-common/src/main/java/com/palantir/gradle/jdks/CurrentOs.java>CurrentOs.java</a>
+ * We cannot depend directly on `gradle-jdks-setup-common` as it might lead to Gradle classLoader issues.
+ */
```
Grim these four files need copying. I think we might be able to shade the dep they come from instead to avoid the copying using gradle-shadow-jar?

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+                .resolve("gradle-jdks");
+    }
+
+    // keep in sync with com.palantir.gradle.jdks.JdksPlugin#isGradleJdkSetupEnabled(org.gradle.api.Project)
```
Might be able to extract this out to another gradle project and shade too.

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Override
+    public void apply(Settings settings) {
```
Would put the `apply` first in the class, reading it just flows a bit better this way.

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+            // see: https://github.com/gradle/gradle/blob/4bd1b3d3fc3f31db5a26eecb416a165b8cc36082/subprojects/core-api/
+            // src/main/java/org/gradle/api/internal/properties/GradleProperties.java#L28
+            if (method.getName().equals("find")) {
+                if (args.length == 1 && args[0].equals("org.gradle.java.installations.auto-detect")) {
```
Might as well extract out the `args.length == 1` to the outer `if`.

CRogers replied:
Same extracting `args[0]` to a variable like `String onlyArg`.

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+                        Files.readString(gradleJdkConfigurationPath).trim();
+                Path installationPath = installationDirectory.resolve(localFilename);
+                if (!Files.exists(installationPath)) {
+                    logger.warn("Could not find toolchain at {}", installationPath);
```
Feels like this should probably error rather than just silently fail?

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+            Os os = CurrentOs.get();
+            Arch arch = CurrentArch.get();
+            try (Stream<Path> stream = Files.list(gradleJdksLocalDirectory).filter(Files::isDirectory)) {
+                return stream.map(path -> path.resolve(String.format("%s/%s/local-path", os, arch)))
```
Might as well just do `path.resolve(os).resolve(arch).resolve("local-path")` rather than string formatting.

CRogers commented on gradle-jdks-settings/src/main/java/com/palantir/gradle/jdks/settings/PatchToolchainJdkPlugin.java:
```diff
@@ -0,0 +1,203 @@
+    public static boolean isGradleJdkSetupEnabled(Path projectDir) {
+        if (GradleVersion.current().compareTo(GradleVersion.version("7.6")) < 0) {
+            throw new RuntimeException("Failed to apply com.palantir.jdks.settings plugin. Gradle JDK setup is enabled"
+                    + " (palantir.jdk.setup.enabled=true) but the version of Gradle is less than 7.6. Please"
+                    + " upgrade to Gradle 7.6 or higher to use this functionality.");
```
Just for my understanding (I think it's fine to not support <7.6), is this due to some internal code change around 7.6?

crogoz replied:
I tried testing with different Gradle Versions (https://github.com/palantir/gradle-jdks/pull/360/files#diff-98ba650ac01c560cdf5ffae45e4a36b0f4786c3956183b2d9cc1d9759b3061ceR87) but anything below Gradle `v7.6` it would fail with `Unsupported class file major version`.

crogoz committed:
run when executing tasks as well

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,252 @@
+
+
+### Running a Gradle build from inside `Intellij`
+`Intelij` doesn't use the `./gradlew` script, instead it uses the [Gradle Tooling API](https://docs.gradle.org/current/userguide/third_party_integration.html#sec:embedding_introduction).
+In the Intellij plugin `palantir-gradle-jdks` a postStartupActivity will run that will do the following:
+* runs `./gradlew javaToolchains` which triggers the installation of the JDKs and the certs (see above) and it will also call the `checkGradleJdkConfigs` task to make sure the Gradle JDK setup was correctly set-up.
+* sets the Gradle JVM version to the Gradle local java home (`GRADLE_LOCAL_JAVA_HOME` which is resolved from the `gradle/gradle-daemon-jdk-version` file).
```
This probably needs an update.

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,252 @@
+## ToolchainsPlugin tasks
+
+The new workflow is set up by [ToolchainsPlugin](../gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java) which gets applied if `palantir.jdk.setup.enabled=true`.
+The plugin won't apply anymore the `baseline-java-versions` plugin, allowing for the configuration of the Java Toolchains as described in the [Gradle docs ](https://docs.gradle.org/current/userguide/toolchains.html)
```
> apply anymore the `baseline-java-versions` plugin
apply the `baseline-java-version` plugin anymore

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,252 @@
+The plugin won't apply anymore the `baseline-java-versions` plugin, allowing for the configuration of the Java Toolchains as described in the [Gradle docs ](https://docs.gradle.org/current/userguide/toolchains.html)
+
+The plugin registers the following tasks:
+- `wrapperJdkPatcher` - finalizes the `wrapper` task, such that everytime the `gradle-wrapper.jar` and/or `./gradlew` files are updated, we will also patch them
```
We don't patch the jar anymore right?

CRogers commented on gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/FileUtils.java:
```diff
@@ -55,5 +55,15 @@ public FileVisitResult postVisitDirectory(Path _dir, IOException exc) throws IOE
+    public static void createDirectories(Path directoryPath) {
+        try {
+            if (!Files.exists(directoryPath)) {
+                Files.createDirectories(directoryPath);
```
Doesn't `createDirectories` not care if the dir already exists?

CRogers commented on gradle-jdks-setup/src/main/resources/gradle-jdks-setup.sh:
```diff
@@ -129,36 +125,55 @@ for dir in "$APP_GRADLE_DIR"/jdks/*/; do
   jdk_installation_directory="$GRADLE_JDKS_HOME"/"$distribution_local_path"
   if [ -d "$jdk_installation_directory" ]; then
-    echo "Distribution $distribution_url already exists in $jdk_installation_directory"
+    echo "Distribution '$distribution_url' already exists in '$jdk_installation_directory'"
```
I think we'll want to hide any output in the case where everything is already set up to avoid noise - at least in the `./gradlew` case (IntelliJ will go to it's own tool window so less of a problem).

CRogers commented on gradle-jdks/build.gradle:
```diff
@@ -44,21 +46,26 @@ gradlePlugin {
+            tags.addAll("java", "jdks")
+        }
+        jdksIdea {
+            id = 'com.palantir.jdks-idea'
```
Probably call this `com.palantir.jdks.idea`, using dots is the style we already use (like `com.palantir.jdks.latest` or `com.palantir.jdks.internal.xxxx`).

CRogers commented on gradle-jdks/src/main/groovy/com/palantir/gradle/jdks/PalantirGradleJdkIdeaPlugin.java:
```diff
@@ -0,0 +1,116 @@
+import org.gradle.plugins.ide.idea.model.IdeaModel;
+import org.xml.sax.SAXException;
+
+public class PalantirGradleJdkIdeaPlugin implements Plugin<Project> {
```
Keep Java files in java source dirs (this is currently in groovy) otherwise they don't get errorprones. If this makes it difficult to access the groovy class you need (due to the silly source set ordering thing), move the groovy class to another project and depend on it, something like `gradle-jdks-groovy`.

crogoz committed:
address comments

crogoz committed:
only log if work by gradlew is done

crogoz committed:
/

crogoz committed:
Merge branch 'develop' into cr/patchin2-generate-gradle

crogoz committed:
move & cleanup

crogoz committed:
Merge branch 'cr/patchin2-generate-gradle' of github.com:palantir/gradle-jdks into cr/patchin2-generate-gradle

crogoz committed:
fix throw & cleanup

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleJdksConfigs.java:
```diff
@@ -0,0 +1,122 @@
+
+    abstract Directory gradleDirectory();
+
+    public abstract void applyGradleJdkFileAction(
+            Path downloadUrlPath, Path localUrlPath, JdkDistributionConfig jdkDistributionConfig);
+
+    public abstract void applyGradleJdkDaemonVersionAction(Path gradleJdkDaemonVersion);
+
+    public abstract void applyGradleJdkJarAction(File gradleJdkJarFile, String resourceName);
+
+    public abstract void applyGradleJdkScriptAction(File gradleJdkScriptFile, String resourceName);
+
+    public abstract void applyCertAction(File certFile, String alias, String content);
```
These should probably be `protected`.

crogoz committed:
fixes

CRogers commented on gradle-jdks-settings/build.gradle:
```diff
@@ -0,0 +1,59 @@
+// workaround, we are removing the artifact from the pom file.
+publishing.publications {
+    maven(MavenPublication) {
+        pom.withXml {
+            var projectDependencyNodes = []
+            asNode().depthFirst { node ->
+                if (node.name().localPart == 'dependency' && node.get("groupId").text() == "com.palantir.gradle.jdks" && node.get("artifactId").text() == "gradle-jdks-root") {
+                    projectDependencyNodes << node
+                }
+            }
+            projectDependencyNodes.each { node ->
+                node.parent().remove(node)
+            }
+        }
```
I'd extract this out to a closure or do something like

```
['pom', 'mavenJava'].each { publicationName ->
    withType(MavenPublication).named(publicationName).configure {
        // xmlMunging
    }
}

crogoz replied:
I ended up doing https://github.com/palantir/gradle-jdks/pull/360/files#diff-81e86c983f0ef998e9aaaada848473136cf165c622939ccf15a2012c1833ba0cR34 as I was getting publicationName doesn't exist for "pluginMaven" (https://github.com/palantir/gradle-jdks/pull/360/commits/50195c9738640b0c7487b23365708af151a8bd51#diff-81e86c983f0ef998e9aaaada848473136cf165c622939ccf15a2012c1833ba0cL46)

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/PalantirGradleJdksIdeaPlugin.java:
```diff
@@ -0,0 +1,116 @@
+import org.gradle.plugins.ide.idea.model.IdeaModel;
+import org.xml.sax.SAXException;
+
+public class PalantirGradleJdksIdeaPlugin implements Plugin<Project> {
```
I wonder if we should only do this when the feature flag is enabled (tbf this might happen, I just can't see where it's used)? Otherwise everyone will install the intellij plugin rather than being able to do a staged rollout.

CRogers replied:
Ah, see it's applied in ToolchainsPlugin

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/PalantirGradleJdksIdeaPlugin.java:
```diff
@@ -0,0 +1,116 @@
+                rootProject == rootProject.getRootProject(),
+                "May only apply PalantirGradleJdkProviderPlugin to the root project");
+        rootProject.getPluginManager().withPlugin("idea", ideaPlugin -> {
+            configureLegacyIdea(rootProject);
```
Can remove legacy idea, this plugin won't even work for them as it works exclusively with the native intellij gradle integration.

CRogers commented on gradle-jdks/src/main/java/com/palantir/gradle/jdks/ToolchainsPlugin.java:
```diff
@@ -0,0 +1,147 @@
+                    GradleJdksEnablement.MINIMUM_SUPPORTED_GRADLE_VERSION));
+        }
+        rootProject.getPluginManager().apply(LifecycleBasePlugin.class);
+        rootProject.getPluginManager().apply("com.palantir.jdks.idea");
```
I'd apply this as a class, helps finding out where it is used.

CRogers commented on idea-plugin/build.gradle:
```diff
@@ -0,0 +1,82 @@
+patchPluginXml {
+    pluginDescription = "Configures the Gradle JDKs for the Toolchains used for a project and for the Gradle Daemon."
+    version = project.version
+    sinceBuild = '232' // TODO: test against this version of IntelliJ to ensure no regressions
```
Is this just copied from PJF? Not sure how we actually work out which build is the earliest we support. I know the plugin runs through compatibility tests when you publish it though.

crogoz replied:
> Is this just copied from PJF?

Yes, I tested this version via `runIde` though by changing the `intellij.version` to 2023.2
I guess most people are running on the latest so I can change this to match `2024.1`

CRogers replied:
We probably want to keep it relatively old so people don't need to upgrade.

> I guess most people are running on the latest so I can change this to match 2024.1

I want this to be true but I keep seeing people on really, really old versions lol

CRogers commented on gradle-jdks-setup/Intellij Gradle JDK.png:
Do we need this anymore? Ideally devs should never have to worry about this with the new listener approach.

CRogers replied:
Actually, this makes sense given how it is referenced in the readme!

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,245 @@
+
+
+### Running a Gradle build from inside `Intellij`
+`Intelij` doesn't use the `./gradlew` script, instead it uses the [Gradle Tooling API](https://docs.gradle.org/current/userguide/third_party_integration.html#sec:embedding_introduction).
```
IntelliJ

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,245 @@
+
+### Running a Gradle build from inside `Intellij`
+`Intelij` doesn't use the `./gradlew` script, instead it uses the [Gradle Tooling API](https://docs.gradle.org/current/userguide/third_party_integration.html#sec:embedding_introduction).
+In the Intellij plugin `palantir-gradle-jdks` an ExternalSystemTaskNotificationListener will be registered that for every `Gradle` task of type `RESOLVE_PROJECT` or `EXECUTE_TASK` will:
```
I wouldn't go into detail about the internals (likely to go out of date eventually, people can just read the code if they want to deeply understand it), just say that before and Gradle task or refresh is run:

CRogers commented on gradle-jdks-setup/README.md:
```diff
@@ -0,0 +1,245 @@
+- `generateGradleJdkConfigs` - generates the [`gradle/` configurations](#gradle-jdk-configuration-directory-structure) required for running the JDKs setup
+- `checkGradleJdkConfigs` - checks that all the `gradle/` configurations are up-to-date. E.g. if the `jdks-latest` plugin is updated, we need to make sure the `gradle/jdks` files reflect the jdk versions.
+- `setupJdks` - task that triggers `wrapperJdkPatcher` and `generateGradleJdkConfigs` and runs the patched `./gradlew` script.
+- `ideaSetup` - task that runs: 
+  - the `./gradlew` script to install & configure the required JDKs
+  - the tasks `checkWrapperJdkPatcher` & `checkGradleJdkConfigs` to ensure the Gradle JDK setup is correctly set-up
+  - the task `javaToolchains` to show the configured toolchains.
```
This is gone too.

CRogers commented on gradle-jdks-setup/README.md:
Nice docs!

CRogers commented:
:+1: given comments

crogoz committed:
address comments

crogoz committed:
fix 2023.2 intellij

crogoz committed:
newline

svc-changelog committed:
Add generated changelog entries

svc-autorelease commented:
Released [0.39.0](https://github.com/palantir/gradle-jdks/releases/tag/0.39.0)
